RavenDB-24300 :  GenAI  stats enhancements

aviv86 opened the pull request:
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-24300/GenAI-stats-enhancements

### Additional description

- Create `Usage` object to hold the LLM used tokens
- Track and report `NumberOfContextObjects` and `TotalCachedContexts` also during Transformation and Update
- Add a dedicated scope for the update phase
- report model call and update phase failures
- Fix and unskip flaky stats test
- Add more stats tests


### Type of change

- [x] Bug fix
- [ ] Regression bug fix
- [ ] Optimization
- [ ] New feature

### How risky is the change?

- [x] Low 
- [ ] Moderate 
- [ ] High
- [ ] Not relevant

### Backward compatibility

- [x] Non breaking change
- [ ] Ensured. Please explain how has it been implemented?
- [ ] Breaking change
- [ ] Not relevant

### Is it platform specific issue?

- [ ] Yes. Please list the affected platforms.
- [x] No

### Documentation update

- [ ] This change requires a documentation update. Please mark the issue on YouTrack using `Documentation Required` tag.
- [x] No documentation update is needed 

### Testing by Contributor

- [x] Tests have been added that prove the fix is effective or that the feature works
- [ ] Internal classes added to the test class (e.g. entity or index definition classes) have the lowest possible access modifier (preferable `private`) 
- [ ] It has been verified by manual testing
- [x] Existing tests verify the correct behavior

### Testing by RavenDB QA team

- [ ] This change requires a special QA testing due to possible performance or resources usage implications (CPU, memory, IO). Please mark the issue on YouTrack using `QA Required` tag.
- [x] No special testing by RavenDB QA team is needed

### Is there any existing behavior change of other features due to this change?

- [ ] Yes. Please list the affected features/subsystems and provide appropriate explanation
- [x] No

### UI work

- [x] It requires further work in the Studio. Please mark the issue on YouTrack using `Studio Required` tag.
- [ ] No UI work is needed

karmeli87 commented on src/Raven.Server/Documents/ETL/Providers/AI/GenAi/GenAiBatchPatchCommand.cs:
```diff
@@ -48,63 +53,75 @@ protected override long ExecuteCmd(DocumentsOperationContext context)
+
+                        statsScope.UpdateFailures++;
+                        _statistics.RecordPartialLoadError(msg, item.DocId);
+                        _logger.Log(LogLevel.Warn, msg);
```
```
if(_logger.IsWarnEnabled)
   _logger.Warn(..)
```

karmeli87 replied:
please check how it is done in other places and do the same.
(also fix all other relevant places here)

aviv86 replied:
fixed

Commit:
RavenDB-24300 : address PR comments

ppekrol commented on src/Raven.Server/Documents/ETL/Providers/AI/GenAi/GenAiScriptTransformer.cs:
```diff
@@ -81,6 +82,7 @@ public override IEnumerable<GenAiScriptResult> GetTransformedResults()
 
     public override void Transform(GenAiItem item, GenAiStatsScope stats, EtlProcessState state)
     {
+        _stats = stats.For(EtlOperations.Transform);
```
for performance reasons this should be created in a ctor with start: false and just started here? We are doing that in indexing for example

aviv86 replied:
done

ppekrol commented on src/Raven.Server/Documents/ETL/Providers/AI/GenAi/GenAiTask.cs:
```diff
@@ -292,7 +296,8 @@ Exception HandleItemError(Task<(string Result, AiUsage Usage)> task, GenAiResult
                     Statistics.RecordPartialLoadError(msg, item.DocId);
-                    Logger.Log(LogLevel.Warn, msg);
+                    if (Logger.IsWarnEnabled)
+                        Logger.Log(LogLevel.Warn, msg);
```
Logger.Warn -> lets use a dedicated method

aviv86 replied:
done

ppekrol committed:
RavenDB-24300 added missing IsWarnEnabled check

ppekrol committed:
RavenDB-24300 wrapping transform stats scope in using to stop counting

ppekrol commented:
@kalczur can you take a look at Studio part?

kalczur commented:
@aviv86 ping me on Monday, we will decide where to display what

kalczur committed:
RavenDB-24300 Show GenAI stats on UI

kalczur commented:
![image](https://github.com/user-attachments/assets/cf68e936-f53a-4afc-a9c9-ebb785b65c61)
![image](https://github.com/user-attachments/assets/3c5b1772-b6e1-417e-8084-44f35959b6ea)
![image](https://github.com/user-attachments/assets/e2e7c1bc-22b0-4a97-9b0a-0d8a45c5250e)
![image](https://github.com/user-attachments/assets/7ba1c1e5-dc2d-439e-8031-b593e449450b)

karmeli87 committed:
RavenDB-24300 : GenAI stats enhancements
- Create Usage object to hold the LLM used tokens
- track and report NumberOfContextObjects and TotalCachedContexts also during Transformation and Update
- add a dedicated scope for the update phase
- report model call and update phase failures
- fix and unskip flaky stats test
- add more stats tests
